mlflow/pyspark/ml/__init__.py

```diff
@@ -107,24 +120,129 @@ def _get_estimator_info_tags(estimator):
     }
 
 
-def _get_pipeline_stage_hierarchy(pipeline):
+def _is_parameter_search_estimator(instance):
+    from pyspark.ml.tuning import CrossValidator, TrainValidationSplit
+    return isinstance(instance, (CrossValidator, TrainValidationSplit))
+
+
+def _is_parameter_search_model(instance):
+    from pyspark.ml.tuning import CrossValidatorModel, TrainValidationSplitModel
+    return isinstance(instance, (CrossValidatorModel, TrainValidationSplitModel))
+
+
+def _should_log_hierarchy(estimator):
+    from pyspark.ml import Pipeline
+    from pyspark.ml.classification import OneVsRest
+    return isinstance(estimator, (Pipeline, OneVsRest)) or \
+        _is_parameter_search_estimator(estimator)
+
+
+AutologgingEstimatorMetadata = namedtuple('AutologgingEstimatorMetadata', [
+    'hierarchy', 'uid2name_map', 'param_search_estimators'
+])
+
+
+def _gen_instance_metadata_recursively(
+        instance,
```

Comment:
Nit: This method fetches metadata for a particular estimator stage, so I think it might be clearer to rename it:

```suggestion
def _gen_stage_metadata_recursively(
        stage,
```